### ISSUE TITLE
blackbox exporter fails to fall back on ip protocol correctly

### ISSUE THREAD

**Issue by fredwangwang** [prometheus/blackbox_exporter#594]
### Host operating system: output of `uname -a`

`Linux 539278d9-ce66-4533-a70a-957a286691ce 4.15.0-74-generic #83~16.04.1-Ubuntu SMP Wed Dec 18 04:56:23 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux`

### blackbox_exporter version: output of `blackbox_exporter -version`

```
blackbox_exporter, version 0.16.0 (branch: HEAD, revision: 991f89846ae10db22a3933356a7d196642fcb9a9)
  build user:       root@64f600555645
  build date:       20191111-16:27:24
  go version:       go1.13.4
```

### What is the blackbox.yml module config.

```
modules:
  http_2xx:
    prober: http
  http_post_2xx:
    prober: http
    http:
      method: POST
  tcp_connect:
    prober: tcp
```

### What logging output did you get from adding `&debug=true` to the probe URL?

```
# curl 'http://127.0.0.1:9115/probe?module=http_2xx&target=google.com&debug=true'
Logs for the probe:
ts=2020-04-07T04:06:53.374292987Z caller=main.go:304 module=http_2xx target=google.com level=info msg="Beginning probe" probe=http timeout_seconds=119.5
ts=2020-04-07T04:06:53.374493678Z caller=http.go:318 module=http_2xx target=google.com level=info msg="Resolving target address" ip_protocol=ip6
ts=2020-04-07T04:06:53.380505647Z caller=http.go:318 module=http_2xx target=google.com level=info msg="Resolved target address" ip=2607:f8b0:4001:c01::64
ts=2020-04-07T04:06:53.380600698Z caller=client.go:250 module=http_2xx target=google.com level=info msg="Making HTTP request" url=http://[2607:f8b0:4001:c01::64] host=google.com
ts=2020-04-07T04:06:53.381954888Z caller=main.go:119 module=http_2xx target=google.com level=error msg="Error for HTTP request" err="Get http://[2607:f8b0:4001:c01::64]: dial tcp [2607:f8b0:4001:c01::64]:80: socket: address family not supported by protocol"
ts=2020-04-07T04:06:53.381997402Z caller=main.go:119 module=http_2xx target=google.com level=info msg="Response timings for roundtrip" roundtrip=0 start=2020-04-07T04:06:53.380694806Z dnsDone=2020-04-07T04:06:53.380694806Z connectDone=2020-04-07T04:06:53.381877775Z gotConn=0001-01-01T00:00:00Z responseStart=0001-01-01T00:00:00Z end=0001-01-01T00:00:00Z
ts=2020-04-07T04:06:53.382025868Z caller=main.go:304 module=http_2xx target=google.com level=error msg="Probe failed" duration_seconds=0.007647161



Metrics that would have been returned:
# HELP probe_dns_lookup_time_seconds Returns the time taken for probe dns lookup in seconds
# TYPE probe_dns_lookup_time_seconds gauge
probe_dns_lookup_time_seconds 0.006025825
# HELP probe_duration_seconds Returns how long the probe took to complete in seconds
# TYPE probe_duration_seconds gauge
probe_duration_seconds 0.007647161
# HELP probe_failed_due_to_regex Indicates if probe failed due to regex
# TYPE probe_failed_due_to_regex gauge
probe_failed_due_to_regex 0
# HELP probe_http_content_length Length of http content response
# TYPE probe_http_content_length gauge
probe_http_content_length 0
# HELP probe_http_duration_seconds Duration of http request by phase, summed over all redirects
# TYPE probe_http_duration_seconds gauge
probe_http_duration_seconds{phase="connect"} 0
probe_http_duration_seconds{phase="processing"} 0
probe_http_duration_seconds{phase="resolve"} 0.006025825
probe_http_duration_seconds{phase="tls"} 0
probe_http_duration_seconds{phase="transfer"} 0
# HELP probe_http_redirects The number of redirects
# TYPE probe_http_redirects gauge
probe_http_redirects 0
# HELP probe_http_ssl Indicates if SSL was used for the final redirect
# TYPE probe_http_ssl gauge
probe_http_ssl 0
# HELP probe_http_status_code Response HTTP status code
# TYPE probe_http_status_code gauge
probe_http_status_code 0
# HELP probe_http_uncompressed_body_length Length of uncompressed response body
# TYPE probe_http_uncompressed_body_length gauge
probe_http_uncompressed_body_length 0
# HELP probe_http_version Returns the version of HTTP of the probe response
# TYPE probe_http_version gauge
probe_http_version 0
# HELP probe_ip_protocol Specifies whether probe ip protocol is IP4 or IP6
# TYPE probe_ip_protocol gauge
probe_ip_protocol 6
# HELP probe_success Displays whether or not the probe was a success
# TYPE probe_success gauge
probe_success 0



Module configuration:
prober: http
http:
    ip_protocol_fallback: true
tcp:
    ip_protocol_fallback: true
icmp:
    ip_protocol_fallback: true
dns:
    ip_protocol_fallback: true
```

### What did you do that produced an error?

This is a blackbox exporter instance running on a gce instance. It is able to resolve the ipv6 address of the hostname I guess due to the internal implementation of golang dns resolver. However, when it tries to establish a ipv6 connection to the address, it fails because the vm instance does not support ipv6 actually.

This is an issue for anyone running blackbox exporter in only ipv4 capable network and leave the prefered ip version to default.

### What did you expect to see?

It correctly fails back to ipv4 and report `probe_success 1`

### What did you see instead?

It failed to estabilsih ipv6 connection and report `probe_success 0`

**Comment by brian-brazil** [comment]
This is as expected, the fallback is only for if it can't find a v6 address in DNS.

**Comment by fredwangwang** [comment]
Hi @brian-brazil I totally understand that is the way it is implemented right now, let me rephrase it as a user's perspective:

When I deploy blackbox exporter on a ipv4 environment,
and I use the default settings for http probes,
As an operator I want to monitor the health of a target host which supports both ipv4 & ipv6 (e.g. www.example.com)
I expect the probe_success to be 1 if the target is reachable in ipv4 (as my network only supports ipv4).

However, given the current logic, I get the `probe_success 0`, simply because the blackbox exporter not able to determine the correct ip protocol to communicate with the target.

**Comment by brian-brazil** [comment]
This is intended behaviour, and is correctly detecting your broken v6 setup. If you want to only test v5 then you need to configure things accordingly.

**Comment by varunrayen** [comment]
I am facing the exact situation, it would much more helpful if you could provide some example config or more information of configuring things.

**Comment by brian-brazil** [comment]
As I said this is the intended behaviour. If you want to test v4, then configure your module to use v4.
